### PR TITLE
Begin migration to TypeScript

### LINKED ISSUE #295: Begin migration to TypeScript
## Summary
This SDK is currently written in plain JavaScript and we do not provide any TypeScript definitions. It's time to fix this!

## Scope
It will most likely be difficult to migrate the entire library to TypeScript, so let's focus on the commonly used, public parts first, such as the makeTxn functions, Transaction class, and the encoding/decoding functions. After this the Algodv2 class would probably be next, but this will be more difficult because we'll have to also annotate every possible REST API response.

In general we need to:
* Incorporate TypeScript compilation into this package's build steps.
* Add types to files one-by-one, focussing on commonly used user-facing classes/functions first.
* Ensure the compiled JS files and type definitions are included when the package is published.

The TypeScript docs provide some guidance about how to migrate a repo: https://www.typescriptlang.org/docs/handbook/migrating-from-javascript.html

## Urgency
Many users have requested TypeScript support (#204, #245, #265) since it would greatly improve the usability of our library and catch some bugs a lot sooner.

### PR DESCRIPTION
**Pull request by jdtzmn**
This PR incorporates TypeScript compilation into the build steps, and migrates the following files and directories to TypeScript:
- [x] `src/transaction.js`
- [x] `src/makeTxn.js`
- [x] `src/encoding`
- [x] `src/nacl/naclWrapper.js`
- [x] `src/main.js`
- [x] `index.js`

It also includes [an example](https://github.com/algorand/js-algorand-sdk/blob/jacob/typescript-migration/examples/typescript_example.ts) showing how to use TypeScript type definitions.

Closes #295.

### CODE REVIEW

File: src/types/transactions/base.ts [hunk 1]
```diff
@@ -0,0 +1,369 @@
+/* eslint-disable no-shadow,no-unused-vars */
+
+/**
+ * Enum for application transaction types.
+ *
+ * The full list is availabe at https://developer.algorand.org/docs/reference/transactions/
+ */
+export enum TransactionType {
```
Comment (jasonpaulos) [inline src/types/transactions/base.ts:8]: I really like this enum

File: src/utils/utils.ts [hunk 2]
```diff
@@ -18,7 +24,7 @@ const JSONbig = require('json-bigint')({ useNativeBigInt: true, strict: true });
  *
  *   Defaults to "default" if not included.
  */
-function parseJSON(str, options = undefined) {
```
Comment (jasonpaulos) [inline src/utils/utils.ts:21]: Nit: can you remove the type annotations in this function's docstring?

File: examples/typescript_example.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    from: addr,
+    to: receiver,
+    amount: 100000,
+    note,
+    suggestedParams,
+    // try adding another option to the list above by using TypeScript autocomplete (ctrl + space in VSCode)
+  });
+
+  const signedTxn = txn.signTxn(sk);
+  console.log(algosdk.decodeObj(signedTxn));
```
Comment (jasonpaulos) [inline examples/typescript_example.ts:43]: Nit: can you use `algosdk.decodeSignedTransaction` here instead?

**Review by jasonpaulos — CHANGES_REQUESTED** [review]
Great work, I'm excited for this! I left a bunch of comments but I didn't see any major issues.

File: src/encoding/address.ts [hunk 4]
```diff
@@ -93,10 +95,10 @@ function isValidAddress(address) {
 
 /**
  * encodeAddress takes an Algorand address as a Uint8Array and encodes it into a string with checksum.
- * @param {Uint8Array} address a raw Algorand address
- * @returns {string} the address and checksum encoded as a string.
+ * @param address a raw Algorand address
+ * @returns the address and checksum encoded as a string.
  */
-function encodeAddress(address) {
+export function encodeAddress(address: Uint8Array | number[]) {
```
[the remark this answers is not in the document]
Comment (jdtzmn) [reply src/encoding/address.ts:101]: No real reason – I looked at what type the `nacl` library wanted for the `genericHash` function, and came up with a reasonable guess at what types we would want to use internally. If the `encodeAddress` will only ever accept `Uint8Array`s, then I can change it.

File: src/encoding/uint64.ts [hunk 5]
```diff
@@ -36,10 +34,13 @@ function encodeUint64(num) {
  *   * "bigint": The integer will always be decoded as a BigInt.
  *
  *   Defaults to "safe" if not included.
- * @returns {number | bigint} The integer that was encoded in the input data. The return type will
+ * @returns The integer that was encoded in the input data. The return type will
  *   be determined by the parameter decodingMode.
  */
-function decodeUint64(data, decodingMode = 'safe') {
+export function decodeUint64(
```
[the remark this answers is not in the document]
Comment (jdtzmn) [reply src/encoding/uint64.ts:40]: Great suggestion! I've added it.

File: src/transaction.ts [hunk 6]
```diff
@@ -1,335 +1,435 @@
...
+  appForeignAssets?: number[];
+  type?: TransactionType;
+  flatFee: boolean;
+  reKeyTo?: Address;
+  nonParticipation?: boolean;
+  group: undefined;
+
+  constructor(transactionObj: AnyTransaction) {
+    // Create a mutable copy of the transaction object
+    const transaction: AnyTransaction = { ...transactionObj };
```
[the remark this answers is not in the document]
Comment (jdtzmn) [reply src/transaction.ts:165]: Very smart.

File: src/makeTxn.ts [hunk 7]
```diff
@@ -16,25 +41,24 @@ const txnBuilder = require('./transaction');
...
- * @returns {Transaction}
  */
-function makePaymentTxnWithSuggestedParams(
-  from,
-  to,
-  amount,
-  closeRemainderTo,
-  note,
-  suggestedParams,
-  rekeyTo = undefined
```
[the remark this answers is not in the document]
Comment (jdtzmn) [reply src/makeTxn.ts:28]: I (wrongly) assumed that optional types would become optional parameters. I've updated the parameters to be optional.

File: src/encoding/address.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (jasonpaulos) [reply src/encoding/address.ts:99]: Yeah let's change this to `Uint8Array`

File: src/transaction.ts [hunk 8]
```diff
@@ -1,335 +1,435 @@
...
+  appClearProgram: Uint8Array;
+  appArgs?: Uint8Array[];
+  appAccounts?: Address[];
+  appForeignApps?: number[];
+  appForeignAssets?: number[];
+  type?: TransactionType;
+  flatFee: boolean;
+  reKeyTo?: Address;
+  nonParticipation?: boolean;
+  group: undefined;
```
[the remark this answers is not in the document]
Comment (jasonpaulos) [reply src/transaction.ts:161]: Sorry I wasn't specific. I think you should make it `group?: Buffer` here and on `TransactionStorageStructure`. Let me know if that doesn't make sense.

[the remark this answers is not in the document]
Comment (jdtzmn) [reply src/transaction.ts:161]: Ah I see. Fixed!

**Review by jasonpaulos — APPROVED** [review]
Looks good!

**Pull request merged**